api/src/main/java/org/pmiops/workbench/db/dao/UserServiceImpl.java

```diff
@@ -583,56 +590,50 @@ public DbUser syncComplianceTrainingStatusV2(DbUser dbUser, Agent agent)
-          final Timestamp dbCompletionTime =
-              accessModuleService
-                  .getAccessModuleStatus(dbUser, AccessModuleName.RT_COMPLIANCE_TRAINING)
-                  .map(AccessModuleStatus::getCompletionEpochMillis)
-                  .map(Timestamp::new)
-                  .orElse(null);
-
-          if (dbCompletionTime == null) {
-            // The badge was previously invalid and is now valid.
-            newComplianceTrainingCompletionTime = now;
-          } else if (!dbUser
-              .getComplianceTrainingExpirationTime()
-              .equals(Timestamp.from(Instant.ofEpochSecond(complianceBadge.getDateexpire())))) {
-            // The badge was previously valid, but has a new expiration date (and so is a new
-            // training)
-            newComplianceTrainingCompletionTime = now;
-          } else {
-            // The badge status has not changed since the last time the status was synced.
-            newComplianceTrainingCompletionTime = dbCompletionTime;
-          }
-
-          // Always update the expiration time if the training badge is valid
-          newComplianceTrainingExpirationTime =
-              Timestamp.from(Instant.ofEpochSecond(complianceBadge.getDateexpire()));
-        } else {
-          // The current badge is invalid or expired, the training must be completed or retaken.
-          newComplianceTrainingCompletionTime = null;
-          newComplianceTrainingExpirationTime = null;
-        }
-      } else {
-        // There is no record of this person having taken the training.
-        newComplianceTrainingCompletionTime = null;
-        newComplianceTrainingExpirationTime = null;
-      }
+      Map<AccessModuleName, Optional<Timestamp>> completionTimes =
+          BADGE_BY_COMPLIANCE_MODULE.entrySet().stream()
+              .collect(
+                  Collectors.toMap(
+                      e -> e.getKey(),
+                      e -> {
+                        AccessModuleName moduleName = e.getKey();
+                        Optional<BadgeDetailsV2> badge =
+                            Optional.ofNullable(userBadgesByName.get(e.getValue()))
+                                .filter(BadgeDetailsV2::getValid)
+                                .filter(b -> b.getLastissued() != null);
+
+                        if (!badge.isPresent()) {
+                          return Optional.empty();
+                        }
+
+                        Instant badgeTime = Instant.ofEpochSecond(badge.get().getLastissued());
+                        Instant dbCompletionTime =
+                            accessModuleService
+                                .getAccessModuleStatus(dbUser, moduleName)
+                                .map(AccessModuleStatus::getCompletionEpochMillis)
+                                .map(Instant::ofEpochMilli)
+                                .orElse(Instant.EPOCH);
+
+                        if (badgeTime.isAfter(dbCompletionTime)) {
```

Comment:
My expectation is that no users will be immediately updated as a result of this change - though it's possible there are edge cases. lastIssued > completionTime should only be possible for users who have retaken the training module; those users likewise should have been updated by their expiration date updating.

We certainly won't see a batch change at least, since this method is only called online by users.